### PR TITLE
Added search promotion for DDG

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/38692

New `BraveSearchOmniboxDDGBanner` feature is introduced to show DDG promotion banner.
As new banner types are added, p3a metric is also updated.

Screenshot 2024-05-29 214548
Screenshot 2024-05-29 214610

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

`BraveSearchConversion*`

This promotion is disabled by default and will be controlled via griffin.
To test easily w/o griffin, we could use `--enable-features=XXX` cmd args

1. Launch brave with `--enable-features=BraveSearchOmniboxDDGBanner:banner_type/type_DDG_B`
2. Set DDG as a default search provider
3. Type anything in omnibox and check DDG promotion is shown at last
4. Set Google as a default search provider
5. Type anything in omnibox and check there is no promotion
6. Set Brave as a default search provider
7. Type anything in omnibox and check there is no promotion

As this promotion is controlled with new griffin study, user could get existing search promotion study togeter
When DDG promotion and existing promotion are enabled both,
user will see DDG promotion if current search provider is DDG.
And user will see existing promotion if current search provider is non-brave search.
To test easily this scenario, we could use `--enable-features=XXX` cmd args also.

1. Launch brave with `--enable-features=BraveSearchOmniboxBanner:banner_type/type_B,BraveSearchOmniboxDDGBanner:banner_type/type_DDG_B`
2. Set DDG as a default search provider
3. Type anything in omnibox and check DDG promotion is shown
4. Set Google as a default search provider
5. Type anything in omnibox and check previous promotion is shown
6. Set brave and check there is no promotion

### CODE REVIEW

**Review by sangwoo108 — APPROVED** [review]
LGTM 👍

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/brave_search_conversion/p3a.h [hunk 1]
```diff
@@ -22,6 +22,15 @@ inline constexpr char kSearchPromoBannerCHistogramName[] =
...
     "Brave.Search.Promo.BannerD";
+inline constexpr char kSearchPromoDDGBannerAHistogramName[] =
+    "Brave.Search.Promo.DDGBannerA";
+inline constexpr char kSearchPromoDDGBannerBHistogramName[] =
+    "Brave.Search.Promo.DDGBannerB";
+inline constexpr char kSearchPromoDDGBannerCHistogramName[] =
+    "Brave.Search.Promo.DDGBannerC";
+inline constexpr char kSearchPromoDDGBannerDHistogramName[] =
+    "Brave.Search.Promo.DDGBannerD";
+
```
Comment (DJAndries) [inline components/brave_search_conversion/p3a.h:33]: everything looks good, we just need to add these to the metric list here: https://github.com/brave/brave-core/blob/374231672c477bb57a71225b748fb3ddd430b9f0/components/p3a/metric_names.h#L26-L28

Comment (simonhong) [reply components/brave_search_conversion/p3a.h:33]: Thanks updated - https://github.com/brave/brave-core/pull/23889/commits/b96c033bd6d2d004b8815d36a987b5010ad5733c

**Comment by ShivanKaul** [comment]
non-blocking comment: is there a way for the user to dismiss this? what happens if the user doesn't interact with the banner?

**Review by DJAndries — APPROVED** [review]
p3a++

**Comment by simonhong** [comment]
> non-blocking comment: is there a way for the user to dismiss this? what happens if the user doesn't interact with the banner?

If user dismiss by clicking close(`x`) button, this banner will not be shown anymore.
If user doesn't interact, we'll add this banner again.
So, there are two ways for user to not show this banner again. Click `x` button or set default button.
Otherwise, user will see it again.

**Pull request merged**